### PR TITLE
Turn file references in the panel into links

### PR DESCRIPTION
**Pull request by zmb3**
Fixes: #550 

image

The test panel is now a legitimate etch component instead of raw HTML.

### CODE REVIEW

**Comment by zmb3** [comment]
Investigate https://www.npmjs.com/package/react-ansi-style for restoring the color functionality.

**Comment by lloiser** [comment]
fyi: I've tried to implement an atom grammar to display ANSI sequences in atom, see https://github.com/lloiser/language-ansi-styles. Maybe I can polish this a bit and use it here instead of ansi2html or react-ansi-style?

**Comment by zmb3** [comment]
I'm not sure if a grammar would work.
That would mean the panel would have to be an editor wouldn't it?  If
that's the case then I'm not sure how to implement links.

**Comment by zmb3** [comment]
ANSI functionality is restored and file links are working.
Need some specs and a way to make sure we don't get paths to temp files (from cover).

**Comment by zmb3** [comment]
Submitted https://github.com/smallhelm/ansi-style-parser/pull/1 to add support for additional colors.  When this is merged we can add styles for the light versions. 

For example, with this PR and a new style, ginkgo's _light_ red output displays correctly:
```
.go-plus-ansi-lightRed {
   color: lighten(red, 20%);
}
```

**Comment by zmb3** [comment]
This should be ready to go.  @lloiser you always seem to catch my mistakes after I merge, I'd love your review :-)

We still suffer from the filepaths sometimes being wrong due to coverage, but this only happens when there is a build error, not when a test fails (which was the original intent of #550).  I think we can handle that separately.

File: lib/test/test-panel.js [hunk 1]
```diff
@@ -14,6 +19,36 @@ export default class TestPanel extends EtchComponent {
     if (this.props.model) {
       this.props.model.view = this
     }
+
+    this.locationRegex = /([\w-/.\\:]*.go:\d+)/g
```
Comment (lloiser) [inline lib/test/test-panel.js]: this does not capture a `<file>:<line>:<column>` like `foo/bar.go:23:4`
tbh: I haven't seen any output that contained this file syntax, so feel free to ignore this comment

File: spec/test/test-panel-spec.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import TestPanel from './../../lib/test/test-panel'
+
+describe('test panel', () => {
+  let testPanel
+
+  beforeEach(() => {
+    testPanel = new TestPanel()
+  })
+
+  describe('make link', () => {
```
Comment (lloiser) [inline spec/test/test-panel-spec.js]: add two more tests:

_pre- and suffix_
input string: `"prefix foo/bar.go:23 suffix"`
output elements:
* SPAN with text `"prefix "`
* A with text `"foo/bar.go:23"`
* SPAN with text `" suffix"`

_multi line text_
input string: `"--- FAIL: TestFail (0.00s)\n\tbar_test.go:23: Error!\nFAIL"`
output elements:
* SPAN with text `"--- FAIL: TestFail (0.00s)\n\t"`
* A with text `"bar_test.go:23"`
* SPAN with text `": Error!\nFAIL"`

File: styles/ansi.less [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+.go-plus-ansi-lightMagenta {
+  color: lighten(magenta, 20%);
+}
+
+.go-plus-ansi-lightCyan {
+  color: lighten(cyan, 20%);
+}
+
+.go-plus-ansi-lightWhite {
+  color: lighten(white, 20%);
```
Comment (lloiser) [inline styles/ansi.less]: light white -> whiter than white 😆
maybe set normal white to a slightly greyish color and light white to white

File: lib/test/test-panel.js (hunk 1 above)
Comment (zmb3) [inline lib/test/test-panel.js]: going to leave this as is - even with a column we'll still link to the correct line

File: lib/test/test-panel.js [hunk 4]
```diff
@@ -27,10 +62,30 @@ export default class TestPanel extends EtchComponent {
...
-      <div ref='content' className='go-plus-test-panel' scrollTop={this.scrollHeight} style={style} innerHTML={output} />
+      <div ref='content' className='go-plus-test-panel' scrollTop={this.scrollHeight} style={style}>
+        <AnsiStyle text={output} mapText={this.makeLink.bind(this)} />
+      </div>
     )
   }
 
+  linkClicked (text, dir) {
+    const {file, line, column} = parseGoPosition(text)
+    const base = dir || projectPath()
```
Comment (zmb3) [inline lib/test/test-panel.js]: Good catch.  I did mean `this.props.dir`

File: lib/test/test-panel.js [hunk 5]
```diff
@@ -27,10 +62,30 @@ export default class TestPanel extends EtchComponent {
...
+    if (!base) {
+      return
+    }
+
+    const filepath = path.join(base, file)
```
Comment (zmb3) [inline lib/test/test-panel.js]: :+1:

File: lib/test/test-panel.js [hunk 6]
```diff
@@ -27,10 +62,30 @@ export default class TestPanel extends EtchComponent {
...
+    fs.access(filepath, fs.constants.F_OK | fs.constants.R_OK, (err) => {
```
Comment (zmb3) [inline lib/test/test-panel.js]: Agreed not necessary.

File: styles/ansi.less (hunk 3 above)
Comment (zmb3) [inline styles/ansi.less]: Good catch.  This was rather mechanical.

The white code is actually light gray, and "light white" is pure white.

**Comment by zmb3** [comment]
I think everything's happy now, with the exception of:
- one of Lukas comments that I did not make code changes for 
- a flaky travis run (not sure why this happens but it seems to be a regular occurrence)

**Comment by joefitzgerald** [comment]
You can ignore Travis failing on beta0 - it has an issue on linux that is known.

**Comment by joefitzgerald** [comment]
This is awesome 👍 ⚡️ !

**Pull request merged**